### ISSUE TITLE
istio 1.17.1 pulling istio pilot 1.18.0-alpha

### ISSUE THREAD

**Issue by cessien** [istio/istio#44237]
### Bug Description

A month ago I followed the [helm instructions to install istio onto a GKE cluster](https://istio.io/latest/docs/setup/install/helm/). Everything worked swimmingly.

Yesterday afternoon, sidecar injection started failing with the message:
```
Error creating: admission webhook "namespace.sidecar-injector.istio.io" denied the request: failed to run injection template: template: inject:212:12: executing "inject" at <.EstimatedConcurrency>: can't evaluate field EstimatedConcurrency in type *inject.SidecarTemplateData
```

Upon investigation, we noticed the istiod container while released with helm @v1.17.1 was scheduled to use the image `gcr.io/istio-testing/pilot:latest` running at version
```
2023-04-04T18:40:40.411154Z	info	version: 1.18-alpha.f991abdcfdf51fcc965dd42529907615cddc709f-f991abdcfdf51fcc965dd42529907615cddc709f-Clean
```

I believe the following two things are incorrect:
1. The default values.yaml file should not default to pulling the latest image, especially if using a specific chart version ([see values.yaml here](https://github.com/istio/istio/blob/release-1.17/manifests/charts/istio-control/istio-discovery/values.yaml))
2. The default repository (hub?) should resolve to istio-release, not istio-testing

### Version

```prose
istioctl version
client version: 1.17.0
control plane version: 1.17.1
data plane version: 1.16.2 (17 proxies), 1.17.1 (7 proxies)

kubectl version --short
Flag --short has been deprecated, and will be removed in the future. The --short output will become the default.
Client Version: v1.25.6
Kustomize Version: v4.5.7
Server Version: v1.25.7-gke.1000
```

**Comment by cessien** [comment]
For those looking for a solution, I was able to override the `pilot.image` field in Values.yaml inline to `gcr.io/istio-release/pilot:1.17.1` and everything started working again

**Comment by estevaofay** [comment]
Same thing happening for 1.17.2.

I was able to fix it by following @cessien 's instructions

**Comment by howardjohn** [comment]
Do not just override the values. if you are seeing this, you are installing wrong. Please ensure you are installing in properly supported methods; patching it after is going to lead to headaches and other bugs.

https://istio.io/ has a comprehensive set of ways to install Istio. `git clone && helm install ./manifests/blah` is not one of them.

```
helm template --version 1.17.1 istio/istiod
```

Or downloading the .tar.gz from the github release are valid ways.

**Comment by estevaofay** [comment]
I'm using ArgoCD to install helm charts in my repo. My approach is to run a `helm pull` for each istio component and install it in it's namespaces accordingly.
It's ArgoCD's job to run `kubectl apply -f istio-base` into the cluster with its manifests
I modify the `values.yaml` to get the desired istio config.
Why would that be considered wrong?

**Comment by howardjohn** [comment]
can you give more info? seems fine to me
```
$ helm pull istio/istiod
$ tar xzvf istiod-1.17.1.tgz
$ cat istiod/values.yaml | rg tag:
  tag: ""
  tag: 1.17.1
```

**Comment by Agile-BG** [comment]
The issue is easy to reproduce, here are the steps:

1. Install istio via istioctl and do istioctl version check, the result in my case is:
└──> ./istio-1.17.2/bin/istioctl version
  client version: 1.17.2
  control plane version: 1.17.2
  data plane version: 1.17.2 (1 proxies)

2. Install over the helm charts for 1.72.2(I am doing argocd application install):
  - base
  - gateway
  
└──> ./istio-1.17.2/bin/istioctl version
client version: 1.17.2
control plane version: 1.17.2
data plane version: 1.17.2 (3 proxies)
 
3. Install also the helm chart:
  - istiod
└──> ./istio-1.17.2/bin/istioctl version
client version: 1.17.2
control plane version: 1.18-alpha.c7b1007c98efb7753930176578ff4b9e635f7be4
data plane version: 1.17.2 (3 proxies)

Before the following commit(or around it), there was no problem to transition from istioctl installation to the helm installation and it has worked fine for months on my lab k8s:

https://github.com/istio/istio/commit/97baf44e71d48b5d5de025f6fcbcbe66fcc72713

After the above commit the only "hybrid" setup that kind of works is to only install the gateway chart after the istioctl setup -> the moment you add the istiod chart and delete one of your ReplicaSet or/and StatefulSet you will get:

Error creating: admission webhook "namespace.sidecar-injector.istio.io" denied the request: failed to run injection template: template: inject:212:12: executing "inject" at <.EstimatedConcurrency>: can't evaluate field EstimatedConcurrency in type *inject.SidecarTemplateData

**Comment by evanrich** [comment]
> The issue is easy to reproduce, here are the steps:

Thank you just found this after setting up istio for the first time and running into the  exact same issue you had.  adding an override in argocd to set the pilot.image to 1.17.2 worked for me as well.

**Comment by howardjohn** [comment]
> Before the following commit(or around it), there was no problem to use the wrong pilot image and transition from istioctl installation to the helm installation - it has worked fine for months on my lab k8s:

I would personally phrase it that there *was* a big problem, you just didn't know it yet.

> I mean is it intended to make people use the istio-testing images(not a question rly)

No, my intent is that any users should NOT use this image. Hence why I am suggesting if you are seeing it used, you installed "wrong" (or there was a bug in our release process somewhere).

> Install on top the helm chart:

This phrase isn't really precise enough since there are 4 places you can get the helm chart from.

1. Helm repo at https://istio-release.storage.googleapis.com/charts
2. Helm repo at gcr.io/istio-release/charts
3. Helm chart from manifests/ in the released ./tar.gz such as https://github.com/istio/istio/releases/download/1.17.2/istio-1.17.2-linux-amd64.tar.gz
4. `git pull https://github.com/istio/istio/` and using manifests/ --- THIS ONE IS WRONG

**Comment by evanrich** [comment]
fwiw @howardjohn , this occurs with the officailly supported documentation:

https://istio.io/latest/docs/setup/install/helm/

I'm using this with argocd to deploy:
```
---
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: istio-base
  namespace: argocd
  annotations:
    argocd.argoproj.io/sync-wave: "2"
spec:
  project: default
  source:
    repoURL: https://istio-release.storage.googleapis.com/charts
    chart: base
    targetRevision: ~1.17.2
  destination:
    server: "https://kubernetes.default.svc"
    namespace: istio-system
---
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: istiod
  namespace: argocd
  annotations:
    argocd.argoproj.io/sync-wave: "3"
spec:
  syncPolicy:
    automated:
      selfHeal: true
  project: default
  source:
    repoURL: https://istio-release.storage.googleapis.com/charts
    chart: istiod
    targetRevision: ~1.17.2
    helm:
      values: |
        pilot.image: docker.io/istio/pilot:1.17.2
  destination:
    server: "https://kubernetes.default.svc"
    namespace: istio-system
---
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: istio-ingressgateway
  namespace: argocd
  annotations:
    argocd.argoproj.io/sync-wave: "4"
spec:
  syncPolicy:
    automated:
      selfHeal: true
  project: default
  source:
    repoURL: https://istio-release.storage.googleapis.com/charts
    chart: gateway
    targetRevision: ~1.17.2
  destination:
    server: "https://kubernetes.default.svc"
    namespace: istio-ingress
```

So option 1 produces this same problem, even IF he's using option 4 and is the wrong one.

**Comment by howardjohn** [comment]
If (1) produces that it's a big issue. Is it only 1.17.2? https://github.com/istio/istio/issues/44237#issuecomment-1499578780 was me trying and failing to reproduce it myself but that was 1.17.1. On my phone currently so it's hard to check but can later as well

**Comment by evanrich** [comment]
> istiod container while released with helm @v1.17.1 was scheduled to use the image `gcr.io/istio-testing/pilot:latest` running at version

 @howardjohn  I've personally only tried 1.17.2, using the manual override mentioned resolved it for me so I haven't tried any other version I dont know about the OP though

**Comment by evanrich** [comment]
FWIW, `helm show values istio/istiod` and looking at pilot shows this:
```
pilot:
  autoscaleEnabled: true
  autoscaleMin: 1
  autoscaleMax: 5
  replicaCount: 1
  rollingMaxSurge: 100%
  rollingMaxUnavailable: 25%

  hub: ""
  tag: ""
  variant: ""

  # Can be a full hub/image:tag
  image: pilot
```

by default if no tag is provided, docker resorts to "latest" however there is no latest, just some 1.17.2 / sha values (https://hub.docker.com/r/istio/pilot) .   I went back and I looked at my 2 nodes, and see the following:

```
gcr.io/istio-testing/pilot:latest                                                                                                              application/vnd.docker.distribution.manifest.list.v2+json sha256:7fb3972b580af9bca8145ca1539b7204b61395d0f124855aa697822644cfca6b 71.4 MiB  linux/amd64,linux/arm64                                                                                              io.cri-containerd.image=managed
gcr.io/istio-testing/pilot@sha256:7fb3972b580af9bca8145ca1539b7204b61395d0f124855aa697822644cfca6b                 
```

and

```
docker.io/istio/pilot:1.17.2                                                                                                                    application/v                                                                                    io.cri-containerd.image=managed
docker.io/istio/pilot@sha256:04b21dec3b67c429340a03fc8dd2eebef91cd5081a229f7552ecc31a3b5bd403
```

the `docker.io/istio/pilot:1.17.2` is the one currently running that works. however the previous node that istio tried to install on, has a diferent repository: `gcr.io/istio-testing/pilot:latest`

I don't see anything in the values file that would point to gcr.io, however it seems multiple people are having the wrong registry being used, and I doubt it's an ArgoCD thing (even thou @estevaofay is using it as well) because NO other image in my cluster pulls from the wrong repo when deployed via Argo.

**Comment by howardjohn** [comment]
```
$ helm template istio/istiod | rg image: | tail -n1
          image: "docker.io/istio/pilot:1.17.2"

$ helm template --repo https://istio-release.storage.googleapis.com/charts istiod | rg image: | tail -n1
          image: "docker.io/istio/pilot:1.17.2"

$ helm show values --repo https://istio-release.storage.googleapis.com/charts --version 1.17.2 istiod | toJson | gron | rg 'hub|tag|image'
json.global.hub = "docker.io/istio";
json.global.imagePullPolicy = "";
json.global.imagePullSecrets = [];
json.global.proxy.image = "proxyv2";
json.global.proxy_init.image = "proxyv2";
json.global.tag = "1.17.2";
json.pilot.hub = "";
json.pilot.image = "pilot";
json.pilot.tag = "";
```

Like you said I can't find a way to get to gcr.io/istio-testing (aside from `git clone` and using that, which is not supported). I would doubt Argo is doing something weird.

Note on the `show values` part they settings are just split between `global` and `pilot`

**Comment by howardjohn** [comment]
Huh, with Argo I can reproduce it!

```
kubectl create namespace argocd
kubectl apply -n argocd -f https://raw.githubusercontent.com/argoproj/argo-cd/stable/manifests/install.yaml
kubectl patch svc argocd-server -n argocd -p '{"spec": {"type": "LoadBalancer"}}'

argocd login 172.18.0.101

kubectl apply -f manifests/charts/base/crds/crd-all.gen.yaml

kubectl create namespace istio-system
cat <<EOF | kubectl apply -f -
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: istiod
  namespace: argocd
spec:
  syncPolicy:
    automated:
      selfHeal: true
  project: default
  source:
    repoURL: https://istio-release.storage.googleapis.com/charts
    chart: istiod
    targetRevision: ~1.17.2
  destination:
    server: "https://kubernetes.default.svc"
    namespace: istio-system
EOF
```

**Comment by estevaofay** [comment]
@howardjohn something else I might add:

When I create my microservice helm charts I just usually have a pretty barebones structure for helm (created using `helm create` command)

This usually includes a templates folder, `Chart.yaml`, `values.yaml` and so on.

When I do a istio pull via helm like described a couple answer above, the istio helm chart comes with this `files/gen-istio-cluster.yaml`  file.

I found that ArgoCD is applying this file instead of generating manifests through the combination of `templates + values.yaml + Chart.yaml`

So when I updated `values.yaml` to use the correct tag, nothing happened on ArgoCD, but when I changed this `gen-istio-cluster.yaml` file, it did.

This might just be me not knowing Helm charts deeply enough to work with this specific scenario, but this is the first time I've seen this happen.

I thought I'd mention it to give more information to the debuggers

**Comment by howardjohn** [comment]
I figured it out (thanks to https://github.com/istio/istio/issues/44237#issuecomment-1505453120 mostly!). Its getting confused thinking it should use kustomize and not helm, the kustomize refers to the wrong image.

Simply putting a dummy helm value is enough to "fix" it:

```
cat <<EOF | kubectl apply -f -
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: istiod
  namespace: argocd
spec:
  syncPolicy:
    automated:
      selfHeal: true
  project: default
  source:
    repoURL: https://istio-release.storage.googleapis.com/charts
    chart: istiod
    targetRevision: ~1.17.2
    helm:
      values: |
        fake: bar
  destination:
    server: "https://kubernetes.default.svc"
    namespace: istio-system
EOF
```

At surface level this seems like possibly wrong behavior by Argocd. However, also partially on us for publishing the kustomize part. Will dig into it a bit more for next steps.

Thanks for all the info provided here !

**Comment by howardjohn** [comment]
Opened https://github.com/argoproj/argo-cd/issues/13210 against Argo.

For the time being, I recommend users just explicitly set some field under `helm`. Could be anything:

```
  source:
    repoURL: https://istio-release.storage.googleapis.com/charts
    chart: istiod
    targetRevision: ~1.17.2
    helm:
      version: "v3"
```

**Comment by estevaofay** [comment]
Loved this @howardjohn glad I could be of assistance. After reading the ArgoCD thread I understand the problem better!

Just for people reading this, I don't use the pattern described above to apply
```yaml
source:
    repoURL: https://istio-release.storage.googleapis.com/charts
    chart: istiod
    targetRevision: ~1.17.2
```

I use ArgoCD's App of Apps pattern. This is why I "open" up the Helm chart and try to apply it like that. Here's a more detailed explanation of what I mean:

my file tree structure is the following

```bash
.
├── apps
│   ├── bootstrap
│   │   ├── Chart.yaml
│   │   ├── templates
│   │   │   ├── applications.yaml
│   │   │   └── namespaces.yaml
│   │   └── values.yaml
│   └── namespaces
│       ├── istio-system.yaml
│       ├── kustomization.yaml
├── charts
│   ├── istio-base
│   ├── istio-gateway
│   ├── istiod
│   │   ├── Chart.yaml
│   │   ├── README.md
│   │   ├── files
│   │   │   ├── gateway-injection-template.yaml
│   │   │   ├── gen-istio.yaml
│   │   │   ├── grpc-agent.yaml
│   │   │   ├── grpc-simple.yaml
│   │   │   └── injection-template.yaml
│   │   ├── kustomization.yaml
│   │   ├── templates
│   │   │   ├── NOTES.txt
│   │   │   ├── autoscale.yaml
│   │   │   ├── clusterrole.yaml
│   │   │   ├── clusterrolebinding.yaml
│   │   │   ├── configmap-jwks.yaml
│   │   │   ├── configmap.yaml
│   │   │   ├── deployment.yaml
│   │   │   ├── istiod-injector-configmap.yaml
│   │   │   ├── mutatingwebhook.yaml
│   │   │   ├── poddisruptionbudget.yaml
│   │   │   ├── reader-clusterrole.yaml
│   │   │   ├── reader-clusterrolebinding.yaml
│   │   │   ├── revision-tags.yaml
│   │   │   ├── role.yaml
│   │   │   ├── rolebinding.yaml
│   │   │   ├── service.yaml
│   │   │   ├── serviceaccount.yaml
│   │   │   ├── telemetryv2_1.13.yaml
│   │   │   ├── telemetryv2_1.14.yaml
│   │   │   ├── telemetryv2_1.15.yaml
│   │   │   ├── telemetryv2_1.16.yaml
│   │   │   ├── telemetryv2_1.17.yaml
│   │   │   └── validatingwebhookconfiguration.yaml
│   │   └── values.yaml
```

Under `apps/bootstrap/templates/application.yaml`

I have an ArgoCD Application descriptor that loops through all chart folders under `charts` including istio.

Here's how this ArgoCD Application manifest looks like before template rendering

```yaml
{{- range $app := .Values.applications }}
{{- range $microservice := $app.microservices }}
---
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: {{ $microservice.name }}
  namespace: argocd
  annotations:
    argocd.argoproj.io/sync-wave: {{ ($microservice.precedence | quote) | default ("5" | quote) }}    
  finalizers:
    - resources-finalizer.argocd.argoproj.io
spec:
  project: bootstrap
  source:
    repoURL: my-repo.git
    targetRevision: {{ $.Values.targetRevision }}
    path: charts/{{ $microservice.name }}
  destination:
    server: https://kubernetes.default.svc
    namespace: {{ ($microservice.namespace | default $app.namespace) | default "default" }}
  syncPolicy:
    automated:
      prune: {{ $microservice.prune | default true }}
      selfHeal: {{ $microservice.selfHeal | default true }}
{{- end }}
{{- end }}
```

So essentially I'm creating an App for each folder under `charts` (istiod, istio-gateway, istio base all included).
<img width="679" alt="Screenshot 2023-04-12 at 18 42 53" src="https://user-images.githubusercontent.com/13035425/231592532-c0fbca55-a3bc-4e78-add8-793ab9d00daf.png">

Here's the rendered out Application for `istiod` specifically 

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: istiod
  namespace: argocd
  annotations:
    argocd.argoproj.io/sync-wave: "5"    
  finalizers:
    - resources-finalizer.argocd.argoproj.io
spec:
  project: bootstrap
  source:
    repoURL: my-repo.git
    targetRevision: my-branch
    path: charts/istiod
  destination:
    server: https://kubernetes.default.svc
    namespace: istio-system
  syncPolicy:
    automated:
      prune: true
      selfHeal: true
```

So I'm not explicitly setting it as a Helm chart as @howardjohn pointed out in his manifest (wherein the bug lies)  but it still happens through my method which I still think applies

**Comment by howardjohn** [comment]
Ok we have fixed this issue on 1.16+ in the next patch releases (which are not out yet). In the meantime please follow https://github.com/istio/istio/issues/44237#issuecomment-1505911661 to explicitly declare things as "helm".

Thanks everyone!
